src/System.Memory/src/System/Buffers/SequenceReader_search.cs

```diff
@@ -0,0 +1,798 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Buffers
+{
+    public ref partial struct SequenceReader<T> where T : unmanaged, IEquatable<T>
+    {
+        /// <summary>
+        /// Try to read everything up to the given <paramref name="delimiter"/>. The reader will be positioned
+        /// after the first occurence of <paramref name="delimiter"/> when successful.
+        /// </summary>
+        /// <param name="span">The read data, if any.</param>
+        /// <param name="delimiter">The delimiter to look for.</param>
+        /// <returns>True if the <paramref name="delimiter"/> was found.</returns>
+        public bool TryReadTo(out ReadOnlySpan<T> span, T delimiter)
+        {
+            ReadOnlySpan<T> remaining = UnreadSpan;
+            int index = remaining.IndexOf(delimiter);
+
+            if (index != -1)
+            {
+                span = remaining.Slice(0, index);
+                AdvanceCurrentSpan(index + 1);
+                return true;
+            }
+
+            return TryReadToMultisegment(out span, delimiter);
+        }
+
+        private bool TryReadToMultisegment(out ReadOnlySpan<T> span, T delimiter)
+        {
+            return TryReadToSlow(out span, delimiter, advancePastDelimiter: true);
+        }
+
+        public bool TryReadTo(out ReadOnlySpan<T> span, T delimiter, bool advancePastDelimiter)
+        {
+            ReadOnlySpan<T> remaining = UnreadSpan;
+            int index = remaining.IndexOf(delimiter);
+
+            if (index != -1)
+            {
+                span = index == 0 ? default : remaining.Slice(0, index);
+                AdvanceCurrentSpan(index + (advancePastDelimiter ? 1 : 0));
+                return true;
+            }
+
+            return TryReadToSlow(out span, delimiter, advancePastDelimiter);
+        }
+
+        private bool TryReadToSlow(out ReadOnlySpan<T> span, T delimiter, bool advancePastDelimiter)
+        {
+            if (!TryReadToInternal(out ReadOnlySequence<T> sequence, delimiter, advancePastDelimiter, CurrentSpan.Length - CurrentSpanIndex))
+            {
+                span = default;
+                return false;
+            }
+
+            span = sequence.IsSingleSegment ? sequence.First.Span : sequence.ToArray();
```

Comment:
Isn't there a good chance this is going to result in unexpected allocations?

Reply:
It is part of the contract that it will allocate when it crosses segment boundaries. I want to add the `Allocator` pattern as discussed, just didn't want to rush it for preview 1.